Feat/categories main page

ArekChr opened the pull request:
### Details
This PR introduces a new categories page. This is a basic version, which means that the ability to select and unselect categories. This future update will allow for the displaying details, removal, and editing of selected categories.

### Fixed Issues
$ [35705](https://github.com/Expensify/App/issues/35705)
PROPOSAL: https://docs.google.com/document/d/1gk3xqOs7epMbUrSSiX8K7YcqfPLVgqEos0sf-D-GMDA/edit#bookmark=id.qdiriu8utc5u


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console
1. Navigate to the workspace settings as an admin within a paid policy group.
2. Verify that you can access the categories page.
3. Verify that categories are listed.
4. Confirm that you can select all categories by clicking on 'name', and deselect them by clicking on 'name' once more.
5. Access workspace settings as a member with a paid group policy.
6. Replace the sub-path /profile with /categories in the URL.
7. Verify that a 'not found' screen is displayed on the central page.
8. Repeat steps 4 through 6 as both an administrator and a member under a free policy.
9. Confirm that you are unable to access the categories page.

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
10. Upload an image via copy paste
11. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console
1. Navigate to the workspace settings as an admin within a paid policy group.
2. Verify that you can access the categories page.
3. Verify that categories are listed.
4. Confirm that you can select all categories by clicking on 'name', and deselect them by clicking on 'name' once more.
5. Access workspace settings as a member with a paid group policy.
6. Replace the sub-path /profile with /categories in the URL.
7. Verify that a 'not found' screen is displayed on the central page.
8. Repeat steps 4 through 6 as both an administrator and a member under a free policy.
9. Confirm that you are unable to access the categories page.

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message: https://expensify.slack.com/archives/C01GTK53T8Q/p1708529745044689
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->
<img width="537" alt="Zrzut ekranu 2024-02-13 o 12 09 42" src="https://github.com/Expensify/App/assets/24796318/b54d5f96-1944-4dc1-8a29-0da740327c73">

</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->
<img width="342" alt="Zrzut ekranu 2024-02-13 o 12 12 43" src="https://github.com/Expensify/App/assets/24796318/96701e7e-5da7-493f-be38-98a152599f6b">
<img width="342" alt="Zrzut ekranu 2024-02-13 o 12 12 33" src="https://github.com/Expensify/App/assets/24796318/2751155b-1497-4518-b970-46d32a869e62">

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->
<img width="561" alt="Zrzut ekranu 2024-02-13 o 11 42 16" src="https://github.com/Expensify/App/assets/24796318/0d8a08f9-ffec-4b99-943f-f5b3e877a53f">

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->
<img width="375" alt="Zrzut ekranu 2024-02-13 o 12 12 53" src="https://github.com/Expensify/App/assets/24796318/432cc381-5d7c-49d9-abbe-3fb408b5d224">

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

<img width="1494" alt="Zrzut ekranu 2024-02-20 o 12 57 47" src="https://github.com/Expensify/App/assets/24796318/9626da4e-c366-4531-bd6b-ea309c275967">


<img width="1278" alt="Zrzut ekranu 2024-02-20 o 12 53 37" src="https://github.com/Expensify/App/assets/24796318/4e1ce2b5-a7ea-4fd3-916d-dd452032a5f9">

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->
<img width="1502" alt="Zrzut ekranu 2024-02-13 o 11 42 14" src="https://github.com/Expensify/App/assets/24796318/3da2752b-9d62-48bf-ae83-34f1f5ce2560">

</details>

ArekChr committed:
feat: init empty workspace categories page

ArekChr committed:
refactor: add right element to ratio list item, refactor workspace page

ArekChr committed:
refactor: remove prop types

ArekChr committed:
fix: remove rightHandSideComponent

ArekChr committed:
fix: rebase

ArekChr committed:
refactor: remove withWindowDimensions

ArekChr committed:
feat: add prop types

ArekChr committed:
refactor: use policyID instead policy.id

ArekChr committed:
refactor: migrate to typescript

ArekChr committed:
fix: ts params issue

ArekChr committed:
feat: handle go back to workspace if policy not found

rezkiy37 commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,117 @@
+    const {translate} = useLocalize();
+    const [selectedCategories, setSelectedCategories] = useState<string[]>([]);
+
+    const data = useMemo(
```
```suggestion
    const categoryList = useMemo(
```

rezkiy37 commented on src/pages/workspace/categories/withPolicyAccessOrNotFound.tsx:
```diff
@@ -0,0 +1,84 @@
+    policy: OnyxTypes.Policy;
+};
+
+export default function (): <TProps extends WithWorkspaceAccessProps, TRef>(
```
AFAIK, we prefer hooks to hocs. So, can this logic be implemented via a hook?

ArekChr replied:
I think replacing this specific hoc to hook might not be the best option here, it manipulates the display screen using not found or fullscreen loading indicator

rezkiy37 replied:
Ah, got it. It makes sense then. 👍

ArekChr commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,117 @@
+
+    const data = useMemo(
+        () =>
+            Object.values(policyCategories ?? {}).map((value) => ({
```
Sure! we can apply to useMemo also

ArekChr committed:
refactor: cr

ArekChr committed:
feat: add loading when categories empty

ArekChr commented:
@luacmartins, regarding slack [thread](https://callstack-hq.slack.com/archives/C036QM0SLJK/p1707832121603409?thread_ts=1707768509.167959&cid=C036QM0SLJK), I'm thinking about what we should do next because we're waiting for the OpenWorkspaceCategoriesPage API command. Can we keep going with other things, or should we wait and put this PR on hold?

mountiny commented:
@ArekChr i think you can continue with as much as possible, as i explained everything works as expected in case of the primary workspace and that is a usecase for 90+% users they only have one workspace

How is this blocking you

luacmartins reviewed (commented):
Sorry, I'm still ooo but I agree with @mountiny that we shouldn't hold on that

ArekChr commented:
> @ArekChr i think you can continue with as much as possible, as i explained everything works as expected in case of the primary workspace and that is a usecase for 90+% users they only have one workspace
> 
> How is this blocking you

I wasn't sure, thanks for the clear confirmation! Let's move forward

trjExpensify commented:
@rushatgabhane is going to review this PR :)

trjExpensify commented:
Maybe helpful for @rushatgabhane's reference, here is the categories page from the design doc on both web and mobile: 

![image](https://github.com/Expensify/App/assets/16232057/19a09bd8-3848-466c-a0ce-bb8243dc20e3)

mountiny commented:
@ArekChr @burczu I think we should use the same Stylized SelectionList wrapper - `TableList` here https://github.com/Expensify/App/issues/35712#issuecomment-1945851478

rushatgabhane commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+
+type WorkspaceCategoriesPageProps = WorkspaceCategoriesOnyxProps & WithWorkspaceAccessProps & StackScreenProps<CentralPaneNavigatorParamList, typeof SCREENS.WORKSPACE.CATEGORIES>;
+
+function WorkspaceCategoriesPage({policyCategories}: WorkspaceCategoriesPageProps) {
```
Given that this page will be a WIP, should this be under a `beta`?

trjExpensify replied:
As this is only intend to be shown in a `collect` type workspace (not `free`) and nobody apart from us are using NewDot on paid plans yet, I don't necessarily think the work on adding a categories page needs to be aggregated under a beta.

trjExpensify replied:
We do however need to not show `Categories` in the workspace settings of `free` workspaces. They aren't allowed to customise and manage categories on that plan type as mentioned [here](https://github.com/Expensify/App/pull/36324#issuecomment-1945904208).

ArekChr replied:
Updated HOC, which includes redirecting to 404 page for the non-paid policy group

rushatgabhane commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+    };
+
+    if (Object.keys(policyCategories ?? {}).length === 0) {
+        return <FullScreenLoadingIndicator />;
```
If a policy doesn't have any categories, showing a perpetual loader is not a good UX imo.
Just want to confirm if this will be replaced later on

rushatgabhane replied:
An empty table with no rows.

The user can then add categories. cc: @Expensify/design

trjExpensify replied:
Ah yeah, good point. It's somewhat unlikely to run into this, but totally possible.

- a workspace is created with a default set of categories 
- if somebody manually deletes all of the individual category values, the page will be empty but the categories feature still `enabled` until they disable the feature overall

So I think we should use the empty state for that: 

![image](https://github.com/Expensify/App/assets/16232057/8951f0de-f104-4ebd-a0de-5abcd9c3cc8c)

shawnborton replied:
Yup, that makes sense to me!

trjExpensify replied:
Cool, also added a bullet and the mock for this in the categories section of the high level for posterity.

mountiny replied:
I think this is why it would be nice to have the wrapper component so it can display this custom empty state too

ArekChr replied:
Could you share the Figma link for this component? I need access to the assets. Thanks!

ArekChr replied:
Thanks @trjExpensify! I've updated empty state component

rushatgabhane commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+                isSelected: selectedCategories.includes(value.name),
+                rightElement: (
+                    <View style={styles.flexRow}>
+                        <Text style={[styles.disabledText, styles.alignSelfCenter]}>{value.enabled ? translate('workspace.common.enabled') : translate('workspace.common.disabled')}</Text>
```
I think we should have a gap between the category name and status. 

<img width="494" alt="image" src="https://github.com/Expensify/App/assets/29673073/206f3bc0-ea81-4ae7-858f-4f985989c8fc">

ArekChr replied:
@rushatgabhane We will use TableList or refactored SelectionList which will be handled in another PR that will cover this issue. More context here https://github.com/Expensify/App/issues/35712#issuecomment-1945851478

rushatgabhane commented on src/components/SelectionList/RadioListItem.tsx:
```diff
@@ -8,21 +8,24 @@ function RadioListItem({item, showTooltip, textStyles, alternateTextStyles}: Rad
-
-            {!!item.alternateText && (
+        <>
+            <View style={[styles.flex1, styles.alignItemsStart]}>
```
tiny bug: the checkbox should have the hand ✋  cursor style.

https://github.com/Expensify/App/assets/29673073/e005cd04-b43e-40a9-ba63-501afa4b828a

ArekChr replied:
I believe this is a separate issue, as the bug also appears on the members page. Should we address it in the current PR or open a new bug issue for it?

rushatgabhane replied:
ah okay, new bug then :)

rushatgabhane commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+                rightElement: (
+                    <View style={styles.flexRow}>
+                        <Text style={[styles.disabledText, styles.alignSelfCenter]}>{value.enabled ? translate('workspace.common.enabled') : translate('workspace.common.disabled')}</Text>
+                        <View style={styles.p1}>
```
Does the screenshot from [design doc](https://github.com/Expensify/App/pull/36324#issuecomment-1945851412) has more pading between the status and arrow right icon?

<img width="286" alt="image" src="https://github.com/Expensify/App/assets/29673073/d273ab05-f3fd-431a-bc8b-ed9e1842dd0c">

<img width="150" alt="image" src="https://github.com/Expensify/App/assets/29673073/c299ac74-5b00-4c76-9f75-bf0517dd4e06">

ArekChr replied:
fixed, thanks!

mountiny commented on src/pages/workspace/WorkspaceInitialPage.tsx:
```diff
@@ -137,6 +137,12 @@ function WorkspaceInitialPage({policyDraft, policy: policyProp, policyMembers, r
             brickRoadIndicator: hasMembersError ? CONST.BRICK_ROAD_INDICATOR_STATUS.ERROR : undefined,
             routeName: SCREENS.WORKSPACE.MEMBERS,
         },
+        {
+            translationKey: 'workspace.common.categories',
+            icon: Expensicons.Folder,
+            action: singleExecution(waitForNavigate(() => Navigation.navigate(ROUTES.WORKSPACE_CATEGORIES.getRoute(policyID)))),
+            routeName: SCREENS.WORKSPACE.CATEGORIES,
+        },
```
Noting here something that needs to be updated:
- The Categories page will be turned on by default for new collect policies in More Features page
- Thats why we are adding it before theMore features page is implemented
- This needs to be **hidden/ not accessible** for free policies
- Only available for paid group policies

ArekChr replied:
@mountiny Since the more features page has not been implemented yet, should I go with showing this functionality only for beta users and ensure it's available only for paid group policies, as mentioned?

mountiny replied:
@ArekChr I think we can handle it in this issue https://github.com/Expensify/App/issues/36606 where the behaviour has been described I think

luacmartins replied:
In addition to only be shown for collect policies, we only want to show this to admins too, right?

trjExpensify replied:
Correct, only admins can see and manage categories.

trjExpensify commented:
@ArekChr also, just to confirm... the `Categories` page should only visible in the workspace settings of `collect` type workspaces, not `free` type workspaces. From the mockup in the OP, it looks like we're showing it in a `free` workspace.

rushatgabhane commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+    const {translate} = useLocalize();
+    const [selectedCategories, setSelectedCategories] = useState<string[]>([]);
+
+    const categoryList = useMemo<PolicyForList[]>(
```
Just a thought: should we **sort** the list based on status? Enabled categories first, and then disabled.

I think we don't want `disabled` categories to come in the way when looking at this page.

trjExpensify replied:
For now, it was decided to stick with alphabetical. We have future plans to add filters and search etc.

rushatgabhane replied:
thanks for the context 😀

rushatgabhane commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+                title={translate('workspace.common.categories')}
+                shouldShowBackButton={isSmallScreenWidth}
+            />
+            <SelectionList
```
Maybe this is know. The headers should be `Name` and `Status`.

<img width="660" alt="image" src="https://github.com/Expensify/App/assets/29673073/e4377dab-31fc-4540-b487-3bc04e352d5c">

But the header in the PR is `Select all`
<img width="914" alt="image" src="https://github.com/Expensify/App/assets/29673073/82af1813-c54a-40ca-9123-aa4f2eddc198">

ArekChr replied:
Same as in this [comment](https://github.com/Expensify/App/pull/36324#discussion_r1491195529)

mountiny commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+
+type WorkspaceCategoriesPageProps = WorkspaceCategoriesOnyxProps & WithWorkspaceAccessProps & StackScreenProps<CentralPaneNavigatorParamList, typeof SCREENS.WORKSPACE.CATEGORIES>;
+
+function WorkspaceCategoriesPage({policyCategories}: WorkspaceCategoriesPageProps) {
```
That means we probably need to add some redirect here/ show 404 when the policy is not Paid Group policy.

The paid policy that this shows up to needs to have the `policy?.isPolicyExpenseChatEnabled` set to true

trjExpensify replied:
>The paid policy that this shows up to needs to have the policy?.isPolicyExpenseChatEnabled set to true

For my own knowledge, why do we need to do this here? Isn't it enough that the overall workspace settings don't show for a paid plan without that. Like, why do we have to add it to each and every page (knowing we'll then need to remove it later)?

mountiny replied:
> Isn't it enough that the overall workspace settings don't show for a paid plan without that. Like, why do we have to add it to each and every page (knowing we'll then need to remove it later)?

Well I think there will be some Higher Order Component wrapping all the toggle-able pages, but essentially even if you do not have this button in UI you can put the categories URL with policyID of a free policy in and then we need some logic to check that this page is not allowed for free policy

trjExpensify replied:
> you can put the categories URL with policyID of a free policy in and then we need some logic to check that this page is not allowed for free policy

Ah, fair enough.. good point. When `isPolicyExpenseChatEnabled` goes away, what happens? Wouldn't using the policy type be better and future proof?

ArekChr replied:
I've updated HOC that checks `isPolicyExpenseChatEnabled` and redirects 404 for non-paid group policies

mountiny replied:
> When isPolicyExpenseChatEnabled goes away, what happens? Wouldn't using the policy type be better and future proof?

We will keep using it until migration is done, then we can remove it safely

ArekChr committed:
refactor: refactor policyAccess to paidPolicyAccess HOC

ArekChr committed:
fix: chevron padding

ArekChr committed:
fix: paid policy condition

ArekChr committed:
refactor: rename paid policy hoc name & props, move to workspace dir

ArekChr committed:
fix: paid policy condition & add team policy hoc

ArekChr committed:
refactor: admin policy access hoc name wip

ArekChr committed:
fix: change team to paid hoc policy

ArekChr committed:
fix: use isPaidGroupPolicy to conditionaly display categories menu item

ArekChr committed:
fix: typo

rushatgabhane commented:
@ArekChr could you please merge latest main to avoid potential TS errors

luacmartins commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+    const styles = useThemeStyles();
+    const theme = useTheme();
+    const {translate} = useLocalize();
+    const [selectedCategories, setSelectedCategories] = useState<string[]>([]);
```
I wonder if we should keep this list as an object instead of array since the order of selected categories doesn't really matter and that would provide us with O(1) access to check if a category is selected and maybe improve performance.

ArekChr replied:
Great idea! It's a practical approach for performance, especially with a growing list of categories

luacmartins commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+                value: value.name,
+                text: value.name,
+                keyForList: value.name,
+                isSelected: selectedCategories.includes(value.name),
```
See my comment above about maybe storing this list as an object, since some policies might have a large number of categories hundreds/thousands

luacmartins commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,122 @@
+    );
+
+    const toggleCategory = (category: PolicyForList) => {
+        setSelectedCategories((prev) => {
+            if (prev.includes(category.value)) {
+                return prev.filter((item) => item !== category.value);
+            }
+            return [...prev, category.value];
+        });
```
Same here, using an object might make this easier

ArekChr replied:
done!

ArekChr committed:
Merge branch 'main' into feat/categories_main_page

ArekChr committed:
fix: old conflict

ArekChr commented on src/SCREENS.ts:
```diff
@@ -207,9 +207,10 @@ const SCREENS = {
         CURRENCY: 'Workspace_Profile_Currency',
-        DESCRIPTION: 'Workspace_Profile_Description',
         NAME: 'Workspace_Profile_Name',
+        DESCRIPTION: 'Workspace_Description',
```
Added by mistake while resolving conflicts, fixed, thanks!

ArekChr committed:
refactor: use selected categories as an object instead of array

ArekChr committed:
feat: add workspace empty state card & header message

luacmartins commented on src/pages/workspace/categories/WorkspaceCategoriesPage.tsx:
```diff
@@ -0,0 +1,128 @@
+
+WorkspaceCategoriesPage.displayName = 'WorkspaceCategoriesPage';
+
+export default withPaidPolicyAccessOrNotFound()(
+    withAdminPolicyAccessOrNotFound()(
+        withOnyx<WorkspaceCategoriesPageProps, WorkspaceCategoriesOnyxProps>({
+            policyCategories: {
+                key: ({route}) => `${ONYXKEYS.COLLECTION.POLICY_CATEGORIES}${route.params.policyID}`,
+            },
+        })(WorkspaceCategoriesPage),
+    ),
+);
```
I think you can try:

```suggestion
export default compose(
    withPaidPolicyAccessOrNotFound,
    withAdminPolicyAccessOrNotFound,
    withOnyx<WorkspaceCategoriesPageProps, WorkspaceCategoriesOnyxProps>({
        policyCategories: {
            key: ({route}) => `${ONYXKEYS.COLLECTION.POLICY_CATEGORIES}${route.params.policyID}`,
        },
    }),
)(WorkspaceCategoriesPage);
```

You'll also have to import compose

luacmartins replied:
Hmm nvm, I think that crashes the App 😅

luacmartins reviewed (commented):
@ArekChr I made a suggestion for the TS error

ArekChr commented:
@luacmartins I'm waiting for the SelectionList to be merged. Also fixing the HOC related TS error

luacmartins commented:
Any luck with the TS error?

ArekChr committed:
Merge branch 'main' into feat/categories_main_page

ArekChr commented:
@luacmartins Still no luck, asked for help on Slack, more details of the issue [here](https://callstack-hq.slack.com/archives/C0593T50UHG/p1708508793478539)

ArekChr committed:
add categories menu item to protectedCollectPolicyMenuItems

ArekChr committed:
fix: ts error

luacmartins commented:
@ArekChr Nice, seems like we got around that error. We just got the [SelectionList PR](https://github.com/Expensify/App/pull/37000) merged as well, so hopefully that unblocks us and we can continue making progress here

ArekChr committed:
refactor: HOC to wrappers

ArekChr commented:
I've removed HOCs and implemented wrappers that does the same logic, test scenario below

https://github.com/Expensify/App/assets/24796318/531a2175-0cdb-41c6-95bb-44d6715a413e

luacmartins commented on src/pages/workspace/AdminPolicyAccessOrNotFoundWrapper.tsx:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    if (shouldShowNotFoundPage) {
+        return <NotFoundPage onBackButtonPress={() => Navigation.goBack(ROUTES.WORKSPACE_PROFILE.getRoute(props.policyID))} />;
```
Yea, that seems fine to me

ArekChr committed:
Merge branch 'main' into feat/categories_main_page

luacmartins reviewed (changes requested):
@ArekChr looking good, but there are a few style issues that we need to fix:

1. The table header says `Select all`, but it should be `Name` like in the design doc.
2. The table header doesn't have the `Status` header
3. The checkbox to the left of `Select all` doesn't have the same padding as the ones to the left of each category
4. We need to add the correct styles for each category row, e.g. background color, border-radius, etc

rushatgabhane commented:
@luacmartins @ArekChr the PR looks good, and tests well.

nitpick:
5. there should be spacing between the category name and status on **small screen**.


<img width="620" alt="image" src="https://github.com/Expensify/App/assets/29673073/b3e1bfd7-f586-4b31-927a-e96275c79533">

ArekChr commented:
Hi @luacmartins, @rushatgabhane.
@burczu is currently working on the new [ListItem](https://github.com/Expensify/App/issues/35712) component for the members page that will cover these issues

ArekChr committed:
fix: improve es translations

luacmartins commented:
@ArekChr we just merged https://github.com/Expensify/App/pull/37074, can you please incorporate the new ListItem into this PR?

ArekChr committed:
Merge branch 'main' into feat/categories_main_page

ArekChr committed:
feat: utilize tablist

ArekChr committed:
fix: add cursor pointer style to checkbox

ArekChr commented:
I've observed that select all checkbox is too easy to hit, if you tap anywhere on the header, it selects all items. is it ok? 


https://github.com/Expensify/App/assets/24796318/8b0efb7c-494d-48cf-bb33-456ba1cd2aa2

rushatgabhane commented:
yeah i think it's ok for now

ArekChr committed:
refactor: remove unused style utils

trjExpensify commented:
Oh hm, I don't think we want that long term though. Like the individual rows, it should be limited to the select all checkbox being clicked/pressed.

rushatgabhane reviewed (approved):
LGTM

ArekChr commented:
@luacmartins Updated tests steps

luacmartins reviewed (approved):
LGTM, we’ll have to address the checkbox toggling issue in a follow up, but we can tackle that when we update each row to navigate to the RHP.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8055342505) to staging by https://github.com/luacmartins in version: 1.4.44-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8088632252) to production by https://github.com/puneetlath in version: 1.4.44-13 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

alitoshmatov commented on src/pages/workspace/AdminPolicyAccessOrNotFoundWrapper.tsx:
```diff
@@ -0,0 +1,66 @@
+        return <FullscreenLoadingIndicator />;
+    }
+
+    if (shouldShowNotFoundPage) {
```
Not found page is shown in sidebar and main page causing two not found pages, ref: https://github.com/Expensify/App/issues/39092
